src/coprocessor/statistics/fmsketch.rs

```diff
@@ -0,0 +1,181 @@
+// Copyright 2017 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// FIXME: remove following later
+#![allow(dead_code)]
+
+use std::collections::HashSet;
+use std::hash::Hasher;
+
+use fnv::FnvHasher;
+use tipb::analyze;
+use coprocessor::codec::datum;
+use coprocessor::codec::datum::Datum;
+use coprocessor::codec::Result;
+use util::as_slice;
+
+/// `FMSketch` is used to count the number of distinct elements in a `[Datum]`.
+pub struct FMSketch {
+    mask: u64,
+    max_size: usize,
+    hash_set: HashSet<u64>,
+}
+
+impl FMSketch {
+    pub fn new(max_size: usize) -> FMSketch {
+        FMSketch {
+            mask: 0,
+            max_size: max_size,
+            hash_set: HashSet::new(),
+        }
+    }
+
+    pub fn build(values: &[Datum], max_size: usize) -> Result<FMSketch> {
+        let mut s = FMSketch::new(max_size);
+        for value in values {
+            try!(s.insert(value));
+        }
+        Ok(s)
+    }
+
+    pub fn merge(&mut self, rs: &FMSketch) {
+        if self.mask < rs.mask {
+            self.update_mask(rs.mask);
+        }
+        for key in &rs.hash_set {
+            self.insert_hash_value(*key)
+        }
+    }
+
+    // ndv returns the number of distinct elements
+    pub fn ndv(&self) -> u64 {
+        (self.mask + 1) as u64 * (self.hash_set.len() as u64)
+    }
+
+    pub fn into_proto(self) -> analyze::FMSketch {
+        let mut proto = analyze::FMSketch::new();
+        proto.set_mask(self.mask);
+        let hash = self.hash_set.into_iter().collect();
+        proto.set_hashset(hash);
+        proto
+    }
+
+    #[inline]
+    fn update_mask(&mut self, mask: u64) {
+        self.mask = mask;
+        self.hash_set.retain(|&x| x & mask == 0);
+    }
+
+    fn insert_hash_value(&mut self, hash_val: u64) {
+        if (hash_val & self.mask) != 0 {
+            return;
+        }
+        self.hash_set.insert(hash_val);
+        if self.hash_set.len() > self.max_size {
+            let mask = (self.mask << 1) | 1;
+            self.update_mask(mask);
+        }
+    }
+
+    fn insert(&mut self, v: &Datum) -> Result<()> {
+        let bytes = try!(datum::encode_value(as_slice(v)));
```

Comment:
In fact, all of the `datum` will be of type `bytes`, since no decode will happen at coprocessor part, can we remove this `encode`?